packages/babel-parser/src/parser/expression.js

```diff
@@ -949,6 +950,18 @@ export default class ExpressionParser extends LValParser {
           this.match(tt._function) &&
           !this.canInsertSemicolon()
         ) {
+          if (this.state.context.pop() !== tc.functionStatement) {
+            // Since "async" is an identifier and normally identifiers
+            // can't be followed by expression, the tokenizer assumes
+            // that "function" starts a statement.
+            // Fixing it in the tokenizer would mean traking not only the
+            // previous token ("async"), but also the one before to know
+            // if its beforeExpr value.
+            // It's easier and more efficient to adjust the context here.
+            throw new Error("Internal error");
+          }
+          this.state.context.push(tc.functionExpression);
```

Comment:
It should be more efficient to check the last element of `this.state.context` and simply replace it with `tc.functionExpression` without `push` and `pop`.